### PR TITLE
NODE-434 SmartContracts AtomicSwap example

### PR DESCRIPTION
**Pull request by monroid**

### CODE REVIEW

File: it/src/test/scala/com/wavesplatform/it/sync/SmartContractsAtomicSwapSuite.scala [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+
+    val proof  = ByteStr(secretText.getBytes())
+    val signed = unsigned.copy(proofs = Proofs(Seq(proof)))
+    val versionedTransferId =
+      sender.signedBroadcast(signed.json() + ("type" -> JsNumber(TransferTransactionV2.typeId.toInt))).id
+
+    nodes.waitForHeightAriseAndTxPresent(versionedTransferId)
+  }
+
+  private def calcDataFee(data: List[DataEntry[_]]): Long = {
```
Comment (ismagin) [inline it/src/test/scala/com/wavesplatform/it/sync/SmartContractsAtomicSwapSuite.scala]: that's copy-paste?

File: it/src/test/scala/com/wavesplatform/it/sync/SmartContractsAtomicSwapSuite.scala [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+    TypeChecker(dummyTypeCheckerContext, untyped).explicitGet()
+  }
+
+  private val sc2 = {
+    val untyped = Parser(s"""
+    let Alice = extract(addressFromString("${AliceBC2}")).bytes
+    let backToBob = extract(addressFromString("${BobBC2}")).bytes
+    let txRecipient = addressFromRecipient(tx.recipient).bytes
+
+    let txTransfer = if((txRecipient == Alice) && (sha256(tx.proof0) == base58'BN6RTYGWcwektQfSFzH8raYo9awaLgQ7pLyWLQY4S4F5') && (100 >= height)) then true else false
```
Comment (ismagin) [inline it/src/test/scala/com/wavesplatform/it/sync/SmartContractsAtomicSwapSuite.scala]: what is  `BN6RTYGWcwektQfSFzH8raYo9awaLgQ7pLyWLQY4S4F5` ?

File: it/src/test/scala/com/wavesplatform/it/sync/SmartContractsAtomicSwapSuite.scala [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+    val toBobBC2TxId = sender.transfer(sender.address, BobBC2, 10 * transferAmount, fee).id
+    nodes.waitForHeightAriseAndTxPresent(toBobBC2TxId)
+
+    val toSwapBC2TxId = sender.transfer(sender.address, swapBC2, transferAmount, fee).id
+    nodes.waitForHeightAriseAndTxPresent(toSwapBC2TxId)
+  }
+
+  test("step1: create and setup smart contract SC1") {
+    val pkSwapBC1 = PrivateKeyAccount.fromSeed(sender.seed(swapBC1)).right.get
+    val script    = ScriptV1(sc1).explicitGet()
```
Comment (ismagin) [inline it/src/test/scala/com/wavesplatform/it/sync/SmartContractsAtomicSwapSuite.scala]: Soo `pkSwapBC1` becomes open to everyone? And then you put money there? :)

File: it/src/test/scala/com/wavesplatform/it/sync/SmartContractsAtomicSwapSuite.scala [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+
+    nodes.waitForHeightAriseAndTxPresent(setScriptId)
+
+    val swapBC1ScriptInfo = sender.addressScriptInfo(swapBC1)
+
+    swapBC1ScriptInfo.script.isEmpty shouldBe false
+    swapBC1ScriptInfo.scriptText.isEmpty shouldBe false
+  }
+
+  test("step2: make transfer to swapBC1, Bob's audit of DataTransaction with hash, txId and SC1") {
```
Comment (ismagin) [inline it/src/test/scala/com/wavesplatform/it/sync/SmartContractsAtomicSwapSuite.scala]: Why do you need data transaction at all?
Aren't proofs sufficient?

File: it/src/test/scala/com/wavesplatform/it/sync/SmartContractsAtomicSwapSuite.scala [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+  }
+
+  private val sc2 = {
+    val untyped = Parser(s"""
+    let Alice = extract(addressFromString("${AliceBC2}")).bytes
+    let backToBob = extract(addressFromString("${BobBC2}")).bytes
+    let txRecipient = addressFromRecipient(tx.recipient).bytes
+
+    let txTransfer = if((txRecipient == Alice) && (sha256(tx.proof0) == base58'BN6RTYGWcwektQfSFzH8raYo9awaLgQ7pLyWLQY4S4F5') && (100 >= height)) then true else false
+    let backToBobAfterHeight = if ((height >= 101) && (txRecipient == txToBob)) then true else false
```
Comment (ismagin) [inline it/src/test/scala/com/wavesplatform/it/sync/SmartContractsAtomicSwapSuite.scala]: that won't compile

File: it/src/test/scala/com/wavesplatform/it/sync/SmartContractsAtomicSwapSuite.scala (hunk 4 above)
Comment (monroid) [reply it/src/test/scala/com/wavesplatform/it/sync/SmartContractsAtomicSwapSuite.scala]: It is just an example of users data exchange from one network to another

File: it/src/test/scala/com/wavesplatform/it/sync/SmartContractsAtomicSwapSuite.scala (hunk 3 above)
Comment (monroid) [reply it/src/test/scala/com/wavesplatform/it/sync/SmartContractsAtomicSwapSuite.scala]: Nope, transfer to swap address at the next step

File: it/src/test/scala/com/wavesplatform/it/sync/AtomicSwapSmartContractSuite.scala [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+          feeAmount = fee + 0.04.waves,
+          attachment = Array.emptyByteArray
+        )
+        .explicitGet()
+
+    assertBadRequest(sender.signedBroadcast(txToSwapBC1.json() + ("type" -> JsNumber(TransferTransactionV2.typeId.toInt))))
+  }
+
+  test("step5: Bob makes transfer OR Alice takes funds back") {
+    val isBobTakesFunds = Random.nextBoolean()
```
[the remark this answers is not in the document]
Comment (monroid) [reply it/src/test/scala/com/wavesplatform/it/sync/AtomicSwapSmartContractSuite.scala]: Rollback will be supported at dedicated review.

File: it/src/test/scala/com/wavesplatform/it/sync/AtomicSwapSmartContractSuite.scala [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+        .explicitGet()
+    }
+
+    val versionedTransferId =
+      sender.signedBroadcast(signed.json() + ("type" -> JsNumber(TransferTransactionV2.typeId.toInt))).id
+
+    nodes.waitForHeightAriseAndTxPresent(versionedTransferId)
+  }
+
+  protected def calcDataFee(data: List[DataEntry[_]]): Long = {
```
[the remark this answers is not in the document]
Comment (monroid) [reply it/src/test/scala/com/wavesplatform/it/sync/AtomicSwapSmartContractSuite.scala]: Yep, as I said at the previous commit I'll create a refactoring and move this code. Please, read my comments before posting yours.

**Pull request merged**